File: labs/lab2-persistent-queue/python/test_pqueue.py
"""Tests for the persistent queue. Run:  python3 -m unittest test_pqueue -v

The tests are ordered the way you should make them pass:
  A_* basics, B_* the FIFO contract, C_* persistence and sharing.
"""
import random
import unittest
from collections import deque

import pqueue as P


class A_Basics(unittest.TestCase):
    def test_empty_is_empty(self):
        self.assertTrue(P.is_empty(P.EMPTY))

    def test_enqueue_makes_nonempty(self):
        self.assertFalse(P.is_empty(P.enqueue(P.EMPTY, 1)))

    def test_single_roundtrip(self):
        x, q = P.dequeue(P.enqueue(P.EMPTY, 42))
        self.assertEqual(x, 42)
        self.assertTrue(P.is_empty(q))

    def test_peek_does_not_remove(self):
        q = P.from_iter([1, 2])
        self.assertEqual(P.peek(q), 1)
        self.assertEqual(P.peek(q), 1)
        self.assertEqual(P.to_list(q), [1, 2])

    def test_dequeue_empty_raises(self):
        with self.assertRaises(IndexError):
            P.dequeue(P.EMPTY)

    def test_peek_empty_raises(self):
        with self.assertRaises(IndexError):
            P.peek(P.EMPTY)


class B_FifoContract(unittest.TestCase):
    def test_fifo_order(self):
        q = P.from_iter([1, 2, 3, 4, 5])
        out = []
        while not P.is_empty(q):
            x, q = P.dequeue(q)
            out.append(x)
        self.assertEqual(out, [1, 2, 3, 4, 5])

    def test_interleaved(self):
        q = P.from_iter([1, 2])
        x, q = P.dequeue(q)          # -> 1
        q = P.enqueue(q, 3)
        y, q = P.dequeue(q)          # -> 2
        z, q = P.dequeue(q)          # -> 3
        self.assertEqual((x, y, z), (1, 2, 3))

    def test_invariant_front_empty_implies_back_empty(self):
        rng = random.Random(1)
        q = P.EMPTY
        for _ in range(500):
            if rng.random() < 0.6:
                q = P.enqueue(q, rng.randint(0, 99))
            elif not P.is_empty(q):
                _, q = P.dequeue(q)
            front, back = q
            if front is None:
                self.assertIsNone(back, "invariant broken: empty front, non-empty back")

    def test_model_based_random(self):
        """Your queue must agree with a reference model on 2000 random ops."""
        rng = random.Random(2026)
        q, model = P.EMPTY, deque()
        for step in range(2000):
            if rng.random() < 0.55:
                x = rng.randint(0, 999)
                q, _ = P.enqueue(q, x), model.append(x)
            elif model:
                x, q = P.dequeue(q)
                self.assertEqual(x, model.popleft(), f"wrong value at step {step}")
            self.assertEqual(P.to_list(q), list(model), f"wrong contents at step {step}")


class C_Persistence(unittest.TestCase):
    def test_old_version_survives_enqueue(self):
        q1 = P.from_iter([1, 2, 3])
        before = P.to_list(q1)
        P.enqueue(q1, 4)
        self.assertEqual(P.to_list(q1), before, "enqueue modified its argument!")

    def test_old_version_survives_dequeue(self):
        q1 = P.from_iter([1, 2, 3])
        before = P.to_list(q1)
        P.dequeue(q1)
        self.assertEqual(P.to_list(q1), before, "dequeue modified its argument!")

    def test_fan_out(self):
        """One present, several futures: dequeue the SAME version twice."""
        q = P.from_iter([10, 20, 30])
        x1, r1 = P.dequeue(q)
        x2, r2 = P.dequeue(q)
        self.assertEqual(x1, x2)
        self.assertEqual(P.to_list(r1), P.to_list(r2))
        r1b = P.enqueue(r1, 99)          # the two futures now diverge
        self.assertEqual(P.to_list(r1b), [20, 30, 99])
        self.assertEqual(P.to_list(r2), [20, 30])

    def test_structural_sharing(self):
        """Persistence is not copying: enqueue must SHARE the front."""
        q1 = P.from_iter([1, 2, 3])      # front is non-empty
        front1, _ = q1
        q2 = P.enqueue(q1, 4)
        front2, _ = q2
        self.assertIs(front2, front1,
                      "enqueue rebuilt the front instead of sharing it")


if __name__ == "__main__":
    unittest.main()

File: labs/lab2-persistent-queue/ocaml/test/test_pqueue.ml
(* Tests for the persistent queue. Run with:  dune runtest
   Ordered in the sequence in which they should be made to pass:
   basics -> FIFO contract -> persistence. *)

open Pqueue_lib

let check name cond =
  if cond then Printf.printf "  ok   %s\n" name
  else (Printf.printf "  FAIL %s\n" name; exit 1)

(* A reference model: the queue as a plain list, front first.
   Slow and obviously correct — that is the point of a model. *)
let model_enqueue x m = m @ [ x ]
let model_dequeue = function [] -> None | x :: rest -> Some (x, rest)

let () =
  print_endline "basics:";
  check "empty is empty" (Pqueue.is_empty Pqueue.empty);
  check "enqueue makes non-empty"
    (not (Pqueue.is_empty (Pqueue.enqueue 1 Pqueue.empty)));
  check "single round-trip"
    (match Pqueue.dequeue (Pqueue.enqueue 42 Pqueue.empty) with
     | Some (42, q) -> Pqueue.is_empty q
     | _ -> false);
  check "dequeue empty is None" (Pqueue.dequeue Pqueue.empty = None);
  check "peek empty is None" (Pqueue.peek Pqueue.empty = None);

  print_endline "FIFO contract:";
  let q = Pqueue.of_list [ 1; 2; 3; 4; 5 ] in
  check "to_list after of_list" (Pqueue.to_list q = [ 1; 2; 3; 4; 5 ]);
  check "peek is first-in" (Pqueue.peek q = Some 1);
  check "interleaved"
    (let q = Pqueue.of_list [ 1; 2 ] in
     match Pqueue.dequeue q with
     | Some (1, q) -> (
         let q = Pqueue.enqueue 3 q in
         match Pqueue.dequeue q with
         | Some (2, q) -> Pqueue.dequeue q |> Option.map fst = Some 3
         | _ -> false)
     | _ -> false);

  (* 1000 pseudo-random operations against the model. *)
  let seed = ref 2026 in
  let rand n = seed := (!seed * 1103515245) + 12345; abs !seed mod n in
  let rec run step q m =
    if step = 0 then true
    else if Pqueue.to_list q <> m then false
    else if rand 100 < 55 then
      let x = rand 1000 in
      run (step - 1) (Pqueue.enqueue x q) (model_enqueue x m)
    else
      match (Pqueue.dequeue q, model_dequeue m) with
      | None, None -> run (step - 1) q m
      | Some (x, q'), Some (y, m') -> x = y && run (step - 1) q' m'
      | _ -> false
  in
  check "1000 random ops agree with the model" (run 1000 Pqueue.empty []);

  print_endline "persistence:";
  let q1 = Pqueue.of_list [ 1; 2; 3 ] in
  let before = Pqueue.to_list q1 in
  let _ = Pqueue.enqueue 4 q1 in
  check "old version survives enqueue" (Pqueue.to_list q1 = before);
  let _ = Pqueue.dequeue q1 in
  check "old version survives dequeue" (Pqueue.to_list q1 = before);

  (* one present, several futures *)
  let q = Pqueue.of_list [ 10; 20; 30 ] in
  check "fan-out: two futures diverge"
    (match (Pqueue.dequeue q, Pqueue.dequeue q) with
     | Some (x1, r1), Some (x2, r2) ->
         x1 = x2
         && Pqueue.to_list (Pqueue.enqueue 99 r1) = [ 20; 30; 99 ]
         && Pqueue.to_list r2 = [ 20; 30 ]
     | _ -> false);

  print_endline "All tests passed."
